plan/logical_plan_builder.go

```diff
@@ -609,9 +609,17 @@ func (b *planBuilder) buildDistinct(child LogicalPlan, length int) LogicalPlan {
 	return plan4Agg
 }
 
-// joinFieldType finds the type which can carry the given types.
-func joinFieldType(a, b *types.FieldType) *types.FieldType {
+// unionJoinFieldType finds the type which can carry the given types in Union.
+func unionJoinFieldType(a, b *types.FieldType) *types.FieldType {
 	resultTp := types.NewFieldType(types.MergeFieldType(a.Tp, b.Tp))
+	// This logic need be understood with buildProjection4Union logic.
+	if a.Tp == mysql.TypeNewDecimal {
+		// the decimal type could be unsigned only when all the decimals to be united are unsigned
+		resultTp.Flag &= b.Flag & mysql.UnsignedFlag
+	} else {
+		// Other types result will be unsigned only if first child be unsigned.
```

Comment:
Non-decimal results will be unsigned when the first SQL statement result in the union is unsigned.